### PR TITLE
Practice New, This, Pointers, and References

### PR DESCRIPTION
**Pull request by christiannwamba**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -179,133 +135,395 @@ int main()
...
 {
-    float add(float lhs, float rhs);
-    float subtract(float lhs, float rhs);
-    float multiply(float lhs, float rhs);
-    float divide(float lhs, float rhs);
+    float* value;
+    FloatType(float v)
+    {
+        value = new float(v);
+    }
```
Comment (matkatmusic) [inline main.cpp:148]: - [ ] revise, use member initializer list

File: main.cpp [hunk 2]
```diff
@@ -179,133 +135,395 @@ int main()
...
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs, double rhs);
-    double multiply(double lhs, double rhs);
-    double divide(double lhs, double rhs);
+    double* value;
+    DoubleType(double v)
+    {
+        value = new double(v);
+    }
```
Comment (matkatmusic) [inline main.cpp:207]: - [ ] revise, use member initializer list

File: main.cpp [hunk 3]
```diff
@@ -179,133 +135,395 @@ int main()
...
 {
-    int add(int lhs, int rhs);
-    int subtract(int lhs, int rhs);
-    int multiply(int lhs, int rhs);
-    int divide(int lhs, int rhs);
+    int* value;
+    IntType(int v)
+    {
+        value = new int(v);
+    }
```
Comment (matkatmusic) [inline main.cpp:266]: - [ ] revise, use member initializer list

File: main.cpp [hunk 4]
```diff
@@ -179,133 +135,395 @@ int main()
...
     if(rhs == 0) 
     {
-        std::cout << "error, integer division by zero will crash the program!\n";
-        std::cout << "returning lhs\n";
-        return lhs;
+        std::cout << "error: integer division by zero is an error and will crash the program!\n";
+        return *this;
     }
-    return lhs / rhs;
+    *(this->value) /= rhs;
```
Comment (matkatmusic) [inline main.cpp:316]: no need to use `this->`
- [ ] revise

File: main.cpp [hunk 5]
```diff
@@ -179,133 +135,395 @@ int main()
...
+    *(this->value) -= rhs;
+    return *this;
 }
-int IntType::multiply(int lhs, int rhs)
+IntType& IntType::multiply(int rhs)
 {
-    return lhs * rhs;
+    *(this->value) *= rhs;
+    return *this;
 }
```
Comment (matkatmusic) [inline main.cpp:308]: no need to use `this->`
- [ ] revise

File: main.cpp [hunk 6]
```diff
@@ -179,133 +135,395 @@ int main()
...
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::divide(double rhs)
 {
     if(rhs == 0.0) 
     {
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";
+        std::cout << "warning: floating point division by zero!\n";
     }
-    return lhs / rhs;
+    *(this->value) /= rhs;
```
Comment (matkatmusic) [inline main.cpp:256]: no need to use `this->`
- [ ] revise

File: main.cpp [hunk 7]
```diff
@@ -179,133 +135,395 @@ int main()
...
-float FloatType::divide(float lhs, float rhs)
+FloatType& FloatType::divide(float rhs)
 {
     if(rhs == 0.0f)
     {
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";
+        std::cout << "warning: floating point division by zero!\n";
     }
-    return lhs / rhs;
+    *(this->value) /= rhs;
```
Comment (matkatmusic) [inline main.cpp:197]: no need to use `this->`
- [ ] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 8]
```diff
@@ -15,7 +15,21 @@ New/This/Pointers/References conclusion
...
 
+struct A{};
 
+struct HeapA
+{
+    A* a;
+    HeapA()
+    {
+        a = new A();
+    }
```
[the remark this answers is not in the document]
Comment (christiannwamba) [reply main.cpp:26]: Thanks a lot for the code examples.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**